pkg/provisioner/ironic/update_opts.go

```diff
@@ -0,0 +1,145 @@
+package ironic
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/go-logr/logr"
+
+	"github.com/gophercloud/gophercloud/openstack/baremetal/v1/nodes"
+)
+
+type optionsSection string
+
+const (
+	propertiesOpts   optionsSection = "properties"
+	instanceInfoOpts optionsSection = "instance_info"
+)
+
+type optionsData map[string]interface{}
+
+func optionValueEqual(current, value interface{}) bool {
+	switch curVal := current.(type) {
+	case string:
+		if newStr, ok := value.(string); ok {
+			return curVal == newStr
+		}
+	case bool:
+		if newBool, ok := value.(bool); ok {
+			return curVal == newBool
+		}
+	case int:
+		if newInt, ok := value.(int); ok {
+			return curVal == newInt
+		}
+	case []interface{}:
+		// newType could reasonably be either []interface{} or e.g. []string,
+		// so we muct use reflection.
+		newType := reflect.TypeOf(value)
+		switch newType.Kind() {
+		case reflect.Slice, reflect.Array:
+		default:
+			return false
+		}
+		newList := reflect.ValueOf(value)
+		if newList.Len() != len(curVal) {
+			return false
+		}
+		for i, v := range curVal {
+			if !optionValueEqual(newList.Index(i).Interface(), v) {
+				return false
+			}
+		}
+		return true
+	case map[string]interface{}:
+		// newType could reasonably be either map[string]interface{} or
+		// e.g. map[string]string, so we muct use reflection.
+		newType := reflect.TypeOf(value)
+		if newType.Kind() != reflect.Map ||
+			newType.Key().Kind() != reflect.String {
+			return false
+		}
+		newMap := reflect.ValueOf(value)
+		if newMap.Len() != len(curVal) {
+			return false
+		}
+		for k, v := range curVal {
+			newV := newMap.MapIndex(reflect.ValueOf(k))
+			if !(newV.IsValid() && optionValueEqual(newV.Interface(), v)) {
+				return false
+			}
+		}
+		return true
+	}
+	return false
+}
+
+func getUpdateOperation(name string, currentData map[string]interface{}, desiredValue interface{}, path string, log logr.Logger) *nodes.UpdateOperation {
+	current, present := currentData[name]
+	if desiredValue != nil {
+		if !(present && optionValueEqual(current, desiredValue)) {
```

Comment:
I simplified `optionValueEqual()` to avoid comparing scalar types by trying `reflect.DeepEqual()` first, and only falling back to reflection for maps/lists where the type is not `interface{}`.

Reply:
That was really a good idea, as it helps in reducing the complexity.